arch/x86/zefi/zefi.py

```diff
@@ -1,134 +1,158 @@
-
-cf.write("static struct zero_seg zefi_zsegs[] = {\n")
-for s in zero_segs:
-    cf.write("    { 0x%x, %d },\n"
-             % (s[0], s[1]))
-cf.write("};\n\n")
-
-cf.write("static uintptr_t zefi_entry = 0x%xUL;\n" % (entry_addr))
-
-cf.close()
-
-print("\nMetadata header generated.\n")
-
-#
-# Build
-#
-
-# First stage ELF binary.  Flag notes:
-#  + Stack protector is default on some distros and needs library support
-#  + We need pic to enforce that the linker adds no relocations
-#  + UEFI can take interrupts on our stack, so no red zone
-#  + UEFI API assumes 16-bit wchar_t
-cmd = [GCC, "-shared", "-Wall", "-Werror", "-I.",
-       "-fno-stack-protector", "-fpic", "-mno-red-zone", "-fshort-wchar",
-       "-Wl,-nostdlib", "-T", ldscript, "-o", "zefi.elf", cfile]
-print(" ".join(cmd))
-subprocess.run(cmd, check = True)
-
-# Extract the .data segment and append our extra blob
-cmd = [OBJCOPY, "-O", "binary", "-j", ".data", "zefi.elf", "data.dat"]
-print(" ".join(cmd))
-subprocess.run(cmd, check = True)
-
-assert (os.stat("data.dat").st_size % 8) == 0
-df = open("data.dat", "ab")
-df.write(data_blob)
-df.close()
-
-# FIXME: this generates warnings about our unused trash section having to be moved to make room.  Set its address far away...
-subprocess.run([OBJCOPY, "--update-section", ".data=data.dat",
-                "zefi.elf"], check = True)
-
-# Convert it to a PE-COFF DLL.
-cmd = [OBJCOPY, "--target=efi-app-x86_64",
-       "-j", ".text", "-j", ".reloc", "-j", ".data",
-       "zefi.elf", "zephyr.efi"]
-print(" ".join(cmd))
-subprocess.run(cmd, check = True)
-
-print("\nBuild complete; zephyr.efi wrapper binary is ready")
+def verbose(msg):
+    if args.verbose:
+        print(msg)
+
+def build_elf(elf_file):
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+
+    cfile = os.path.join(base_dir, "zefi.c")
+    ldscript = os.path.join(base_dir, "efi.ld")
+
+    assert os.path.isfile(cfile)
+    assert os.path.isfile(ldscript)
+
+    #
+    # Open the ELF file up and find our entry point
+    #
+    fp = open(elf_file, "rb")
+    ef = elftools.elf.elffile.ELFFile(fp)
+
+    symtab = ef.get_section_by_name(".symtab")
+    entry_addr = symtab.get_symbol_by_name(ENTRY_SYM)[0].entry.st_value
+
+    verbose("Entry point address (symbol: %s) 0x%x" % (ENTRY_SYM, entry_addr))
+
+    #
+    # Parse the ELF file and extract segment data
+    #
+
+    data_blob = b''
+    data_segs = []
+    zero_segs = []
+
+    for seg in ef.iter_segments():
+        h = seg.header
+        if h.p_type != "PT_LOAD":
+            continue
+
+        assert h.p_memsz >= h.p_filesz
+        assert (h.p_vaddr % 8) == 0
+        assert len(seg.data()) == h.p_filesz
+
+        if h.p_filesz > 0:
+            sd = seg.data()
+            verbose("%d bytes of data at 0x%x, data offset %d"
+                % (len(sd), h.p_vaddr, len(data_blob)))
+            data_segs.append((h.p_vaddr, len(sd) / 8, len(data_blob) / 8))
```

Comment:
I just realised that `len(sd)` probably matches with `h.p_filesz` since there's the `assert len(seg.data()) == h.p_filesz`, so if that's no longer a multiple of 8 then the above division would leave the last few bytes out.

Reply:
..meaning the second parameter of `data_segs.append()` probably needs to be rounded up, or the left-over bytes should be copied to a separate 8-byte buffer and a second `append()` call done for them. @andyross thoughts?

Reply:
please push a patch that makes those changes to this PR.

Reply:
Yeah.  The reason those asserts were there is that the C implementation of the segment copies is written to use qwords, and I was "pretty sure but not certain" that the ELF output from our linker scripts was going to be padded appropriately, so I threw some asserts in there just for safety.   Obviously that's not the case and we need to pad.

Reply:
@andyross thanks for clarifying. One more question: should `seg.data()` be considered read-only, thereby requiring a separate buffer to copy the overflowing bytes to, or is it ok to directy extend it with the padding?

Reply:
.. Or maybe I'm just confusing myself since you're checking the length of `sd` but copying from `h.p_vaddr`. I'm just not sure where exactly the simplest way to deal with the padding is.

Reply:
@andyross never mind my previous babbling :) I see that what you're appending is just bookkeeping information and the actual copying of data happens later. However the question still remains, since I guess we can't just request to copy extra bytes from an address if the data isn't there (unlike the zero_segs). Would it perhaps make more sense to "fix" the linker scripts so zefi doesn't need to do the padding?

Reply:
It probably would.  This tool has the architecture it does simply because I wanted to eliminate variables.  I knew the existing ELF file, loaded as it directs, works.  So I wanted to duplicate that.  Modifying the linker scripts simultaneously would be changing too many things at once. :)

Reply:
seg.data() is an object owned by the elftools parser.  I wouldn't modify it directly for hygine reasons, but doing so might work.  I have no idea ;)